### DISCUSSION TITLE
Updating constant buffer not working as expected

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#370] [Q&A]
I have a light buffer, as constructed here...
```cs
int lightBufferSize = Marshal.SizeOf(typeof(LightBufferData)); // 32 bytes
ID3D11Buffer cbLight = device.CreateBuffer(lightBufferSize, BindFlags.ConstantBuffer, ResourceUsage.Dynamic, CpuAccessFlags.Write);
```
And I update it like so...
```cs
// Light buffer data
[StructLayout(LayoutKind.Sequential)]
internal struct LightBufferData
{
    public Vector4 Position;
    public Vector4 Color;

    public LightBufferData(Vector3 position, float r, float g, float b)
    {
        Position = new Vector4(position, 1);
        Color = new Vector4(r, g, b, 1);
    }
}

// Update code
ID3D11DeviceContext dc = GraphicsManager.DxDeviceContext;
LightBufferData lightData = new(light.Position, Luminosity, Luminosity, Luminosity);
MappedSubresource cbLightMRes = dc.Map(cbLight, MapMode.WriteDiscard);
Unsafe.Copy(cbLightMRes.DataPointer.ToPointer(), ref lightData);
dc.Unmap(cbLight, 0);
dc.VSSetConstantBuffer(2, cbLight);
```
The shader...
```hlsl
cbuffer LightBuffer : register(b2)
{
    float4 LightPosition;
    float4 LightColor;
}

struct PixelShaderInput
{
    float4 svp : SV_Position;
    float3 pos : TEXCOORD0;
    float3 nor : NORMAL;
    float4 col : COLOR;
    float3 LDir : LightDirection;
    float3 VDir : ViewDirection;
};

float4 main(PixelShaderInput input) : SV_TARGET
{
    float3 color = LightColor.xyz;
    return float4(color, 1 * pow(2, 16));
}
```

This isn't working, the light position is updating, but the light color is not. LightColor values are always 0 in the shader. Can anyone see anything wrong here?

**Comment by Gavin-Williams** [comment]
I stumbled on a trick to get graphics diagnostics working (it hasn't been working for a couple of weeks for me) and I can see that the light buffer is actually updating. 

![Screenshot 2022-12-13 183512](https://user-images.githubusercontent.com/4374176/207254267-7a6dfa2d-99a8-491a-9db6-b3c5b33ed2ff.png)

**Reply by amerkoleci** [reply]
When I need to send data to GPU I always use structure or even readonly struct/readonly record struct and make the fields are 16byte aligned, for example if I have Vector3 field I add another float to pad the structure.

This is common pattern when working with constant buffer data.

**Comment by Gavin-Williams** [comment]
When I stepped through the shader running, even though the buffer is being updated, the shader is showing all zeros for LightColor.

**Reply by amerkoleci** [reply]
I've setup helper stuff you can use in Vortice.Windows.Samples:

https://github.com/amerkoleci/Vortice.Windows.Samples/blob/main/src/Vortice.Framework/D3D11ConstantBuffer.cs

And the example that show how to achieve the stuff you're trying:

https://github.com/amerkoleci/Vortice.Windows.Samples/blob/main/src/Direct3D11/DrawTexturedCube/Program.cs#L109
